### ISSUE TITLE
Add support for Wikipedia and any MediaWiki website

### ISSUE THREAD

**Issue by jthingelstad** [keybase/keybase-issues#704]
I spent some time considering how Keybase.io could support Wikipedia and any MediaWiki hosted website. I think I have found a way that will work even on wide open editable wikis. It’s somewhat complex, so I wrote a blog post describing my proposed solution.

http://thingelstad.com/keybase-and-mediawiki/

Thank you for your consideration! /cc @hexmode @kghbln

**Comment by demon** [comment]
Rather than copy+pasting the permalink to the edit you could probably just validate it by fetching the revision from the API (you'd know the page name already) and making sure it lines up with what's expected (can validate user, sha1, timestamp, etc)

Longtime MW dev here so let us know if we can do anything to make it easier.

**Comment by rastus-vernon** [comment]
@demon
Yes, using the MediaWiki API for this is smarter than fetching HTML pages. Some wikis do not have the API enabled, but it might be better to not support them in order to avoid the additional complexity.

In that case, I think the following would work (replacing `John_Doe` by the user name, and changing the URLs to the right wiki):

1. Load https://en.wikipedia.org/w/api.php?action=query&list=logevents&letype=import&letitle=User:John_Doe/Keybase&format=json and verify that `query.logevents` is empty. If it isn't, the page has been imported and the revision history might have been faked, so we cannot trust this.
2. Find the first revision of the page, at https://en.wikipedia.org/w/api.php?action=query&prop=revisions&titles=User:John_Doe/Keybase&rvlimit=1&rvprop=content|user&rvdir=newer.
3. Verify that the revision comes from the user.
4. Validate the signatures in the message and check that everything is fine.

**Comment by demon** [comment]
I think as long as the revision is authored by the user and it's in the user's space there's not much to worry about. In which case we'd probably just want their latest revision instead of the first.

There's also the possibility that their key changes and they just post the new identify message to validate again rather than deleting and recreating the page (very possible if the user's not an admin) in which case we definitely don't want the first revision.

**Comment by rastus-vernon** [comment]
I didn't think about the case where the key changes. Yes, we definitely want the latest revision. Only caring about the latest revision also solves the problem in part since it means the user can just undo his changes if he realizes he has been manipulated (which is already unlikely to happen).

**Comment by hakamadare** [comment]
how about leveraging Wikipedia's existing [Committed identity](https://en.wikipedia.org/wiki/Wikipedia:Wikipedia_Signpost/2007-05-14/Committed_identity) feature?

**Comment by tgr** [comment]
It would be great to see this implemented. Any way we can help it move forward?

Re: page import, it's pointless to worry about it. Large wikis disable or severely restrict them (Wikipedia disables file imports and only allows direct import from sister wikis which have the same user database), and on small wikis you cannot avoid the risk of someone tampering with the content anyway (e.g. they can import from file and then oversight log events). Just fetch `https://meta.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&prop=revisions&titles=User:<username>/keybase&rvprop=content&rvlimit=1&rvuser=<username>` (or en.wikipedia.org, but meta.wikimedia.org is the closest to a language-independent central hub for Wikipedias) and then grab `.query.pages[0].revisions[0].content` from the JSON response.

**Comment by samtarling** [comment]
Apologies for dragging up an old 'issue', but this would be really great 😄 

@tgr does https://book.keybase.io/guides/proof-integration-guide help at all? Is this something Wikimedia could look at providing for their projects _at least_? Perhaps a tool on [toolforge](https://toolforge.org) (OAuth --> Tool --> Exposed endpoints per linked guide?)

**Comment by kghbln** [comment]
The original post from Thingles is now here: https://www.thingelstad.com/2014/05/11/keybase-and-mediawiki.html

**Comment by samtarling** [comment]
> 
> 
> It would be great to see this implemented. Any way we can help it move forward?
> 
> Re: page import, it's pointless to worry about it. Large wikis disable or severely restrict them (Wikipedia disables file imports and only allows direct import from sister wikis which have the same user database), and on small wikis you cannot avoid the risk of someone tampering with the content anyway (e.g. they can import from file and then oversight log events). Just fetch `https://meta.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&prop=revisions&titles=User:<username>/keybase&rvprop=content&rvlimit=1&rvuser=<username>` (or en.wikipedia.org, but meta.wikimedia.org is the closest to a language-independent central hub for Wikipedias) and then grab `.query.pages[0].revisions[0].content` from the JSON response.

I've logged this "our" side at https://phabricator.wikimedia.org/T294791 :)
